HIRS_Utils/src/main/java/hirs/tpm/eventlog/uefi/UefiSignatureList.java

```diff
@@ -0,0 +1,204 @@
+ * @throws NoSuchAlgorithmException if there's a problem hashing the certificate.
+ * @throws IOException If there's a problem parsing the signature data.
+ */
+UefiSignatureList(final byte[] list)
+        throws CertificateException, NoSuchAlgorithmException, IOException {
+
+  byte[] guid = new byte[UefiConstants.SIZE_16];
+  System.arraycopy(list, 0, guid, 0, UefiConstants.SIZE_16);
+  signatureType = new UefiGuid(guid);
+
+  byte[] lSize = new byte[UefiConstants.SIZE_4];
+  System.arraycopy(list, UefiConstants.OFFSET_16, lSize, 0, UefiConstants.SIZE_4);
+  listSize = HexUtils.leReverseInt(lSize);
+
+  byte[] hSize = new byte[UefiConstants.SIZE_4];
+  System.arraycopy(list, UefiConstants.OFFSET_20, hSize, 0, UefiConstants.SIZE_4);
+
+  byte[] sSize = new byte[UefiConstants.SIZE_4];
+  System.arraycopy(list, UefiConstants.OFFSET_24, sSize, 0, UefiConstants.SIZE_4);
+  signatureSize = HexUtils.leReverseInt(sSize);
+
+  sigData = new byte[signatureSize];
+  System.arraycopy(list, UefiConstants.OFFSET_28, sigData, 0, signatureSize);
+  processSignatureList(sigData);
+}
+
+/**
+ * EFI Signature list constructor.
+ * @param lists  ByteArrayInputStream containing an EFI Signature list.
+ * @throws IOException If there's a problem in reading he input stream.
+ * @throws CertificateException If there's a problem parsing the X509 certificate.
+ * @throws NoSuchAlgorithmException if there's a problem hashing the certificate.
+ */
+UefiSignatureList(final ByteArrayInputStream lists)
+                  throws IOException, CertificateException, NoSuchAlgorithmException {
+  byte[] guid = new byte[UefiConstants.SIZE_16];
+  lists.read(guid);
+  signatureType = new UefiGuid(guid);
+
+  if (!isValidSigListGUID(signatureType)) {
+      processSignatureData(lists);
+  } else { // valid SigData Processing
+      byte[] lSize = new byte[UefiConstants.SIZE_4];
+      lists.read(lSize);
+      listSize = HexUtils.leReverseInt(lSize);
+
+      byte[] hSize = new byte[UefiConstants.SIZE_4];
+      lists.read(hSize);
+
+      byte[] sSize = new byte[UefiConstants.SIZE_4];
+      lists.read(sSize);
+      signatureSize = listSize - UefiConstants.SIZE_28;
+      sigData = new byte[signatureSize];
+      lists.read(sigData);
+      processSignatureList(sigData);
+  }
+}
+
+/**
+ * Method for processing a set of EFI SignatureList(s).
+ * @param sigData  Byte array holding one or more SignatureLists
+ * @throws CertificateException If there's a problem parsing the X509 certificate.
+ * @throws NoSuchAlgorithmException if there's a problem hashing the certificate.
+ * @throws IOException If there's a problem parsing the signature data.
+ */
+private void processSignatureList(final byte[] efiSigData)
+             throws CertificateException, NoSuchAlgorithmException, IOException {
+  efiSigDataIS = new ByteArrayInputStream(efiSigData);
+  while (efiSigDataIS.available() > 0) {
+    UefiSignatureData tmpSigData = new UefiSignatureData(efiSigDataIS, signatureType);
```

Comment:
For performance, just start the variable tpmSigData outside of the while loop rather than creating a new instance every single time.

Reply:
This creates the UEFISignatureData only if there data available.